synapse/handlers/register.py

```diff
@@ -128,6 +128,15 @@ def check_username(self, localpart, guest_access_token=None, assigned_user_id=No
                     errcode=Codes.FORBIDDEN,
                 )
 
+        if guest_access_token is None:
+            try:
+                int(localpart)
+                raise SynapseError(
+                    400, "Numeric user IDs are reserved for guest users."
```

Comment:
Could be nice to send errcode `M_INVALID_USERNAME`, for the moment, the client receives `M_UNKNOWN`:

```json
{"errcode":"M_UNKNOWN","error":"Numeric user IDs are reserved for guest users."}
```